designs/2019-expose-rules-to-formatters/readme.MD

```diff
@@ -0,0 +1,159 @@
+
+```js
+var rulesViolated = {};
+...
+function printRules() {
+    var lines = "*** RULES:\n";
+    rulesViolated.forEach(function (ruleMetadata, ruleId) {
+        lines += ruleId;
+        
+        if (ruleMetadata.docs.description) {
+            lines += ": " + ruleMetadata.docs.description;
+        }
+        
+        lines += "\n";
+        
+        if (ruleMetadata.docs.url) {
+            lines += ruleMetadata.docs.url + "\n";
+        }
+    });
+    return lines;
+}
+```
+
+## Drawbacks
+
+This is a fairly innocuous change in that it is additive, non-breaking (mostly, see Backwards Compatibility), and does not change any of ESLint's core functionality. A downside is that we will be exposing the Rule data model to third-party developers, so future changes could break existing formatters. For example, removing or renaming an existing property, or changing the structure of the Rule.meta object.
+
+## Backwards Compatibility Analysis
+
+Since this change is manifested as a new argument to the formatter's exported interface function, existing formatter code that implements the exported interface function will not be affected and will continue to function even without adding the new argument to their exported function.
+
+(The following paragraph also appears in the Formatters section.)
+A formatter that assigns a function reference to the exported interface function could exhibit unexpected behavior depending on the signature of the referenced function. For example, since this change's second argument is a complex object, a referenced function that expects a Number as its second argument could cause an exception.
+
+## Alternatives
+
+<!--
+    What other designs did you consider? Why did you decide against those?
+
+    This section should also include prior art, such as whether similar
+    projects have already implemented a similar feature.
+-->
+* Including the rule metadata in the result object. This approach results in redundant data being returned, and includes external metadata properties that are not directly relevant.
+* Pass the rules map itself as a argument to the formatter's exported interface function. This approach makes it messier to add additional data in the future, since new arguments would be necessary.
+
+## Open Questions
+
+<!--
+    This section is optional, but is suggested for a first draft.
+
+    What parts of this proposal are you unclear about? What do you
+    need to know before you can finalize this RFC?
+
+    List the questions that you'd like reviewers to focus on. When
+    you've received the answers and updated the design to reflect them, 
+    you can remove this section.
+-->
+* Is it possible for a formatter to be invoked even though no rules have been run? IOW, could the caller suppress the inbox rules without providing any custom rules? The rules collection would be empty in this case, which formatters could potentially mishandle.
```

Comment:
Yes. It's possible to run ESLint without any rules turned on and only return syntax errors.